clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -0,0 +1,168 @@
+        this.state = MemberState.UNJOINED;
+        this.assignorSelection = AssignorSelection.defaultAssignor();
+        this.targetAssignments = new ArrayList<>();
+    }
+
+    public MembershipManagerImpl(String groupId, String groupInstanceId, AssignorSelection assignorSelection) {
+        this(groupId);
+        this.groupInstanceId = Optional.ofNullable(groupInstanceId);
+        setAssignorSelection(assignorSelection);
+    }
+
+    /**
+     * Update assignor selection for the member.
+     *
+     * @param assignorSelection New assignor selection
+     * @throws IllegalArgumentException If the provided assignor selection is null
+     */
+    public void setAssignorSelection(AssignorSelection assignorSelection) {
+        if (assignorSelection == null) {
+            throw new IllegalArgumentException("Assignor selection cannot be null");
+        }
+        this.assignorSelection = assignorSelection;
+    }
+
+    private void transitionTo(MemberState nextState) {
+        if (!nextState.getPreviousValidStates().contains(state)) {
+            // TODO: handle invalid state transition
+            throw new RuntimeException(String.format("Invalid state transition from %s to %s",
+                    state, nextState));
+        }
+        this.state = nextState;
+    }
+
+    @Override
+    public String groupId() {
+        return groupId;
+    }
+
+    @Override
+    public String groupInstanceId() {
+        // TODO: review empty vs null instance id
+        return groupInstanceId.orElse(null);
+    }
+
+    @Override
+    public String memberId() {
+        return memberId;
+    }
+
+    @Override
+    public int memberEpoch() {
+        return memberEpoch;
+    }
+
+    @Override
+    public void updateState(ConsumerGroupHeartbeatResponseData response) {
+        if (response.errorCode() == Errors.NONE.code()) {
+            this.memberId = response.memberId();
+            this.memberEpoch = response.memberEpoch();
+            targetAssignments.add(response.assignment());
+            transitionTo(MemberState.PROCESSING_ASSIGNMENT);
+        } else {
+            if (response.errorCode() == Errors.FENCED_MEMBER_EPOCH.code() || response.errorCode() == Errors.UNKNOWN_MEMBER_ID.code()) {
+                resetMemberIdAndEpoch();
+                transitionTo(MemberState.UNJOINED);
+            } else if (response.errorCode() == Errors.UNRELEASED_INSTANCE_ID.code()) {
+                transitionTo(MemberState.FAILED);
+            }
+        }
+    }
+
+    public void onAssignmentProcessSuccess(ConsumerGroupHeartbeatResponseData.Assignment assignment) {
+        currentAssignment = assignment;
+        targetAssignments.remove(assignment);
+        transitionTo(MemberState.STABLE);
+    }
+
+    public void onAssignmentProcessFailure(Throwable error) {
+        // TODO: handle failure scenario when the member was not able to process the assignment
+    }
+
+    private void resetMemberIdAndEpoch() {
+        this.memberId = "";
```

Comment:
do we ever need to reset memberId?

Reply:
This is used in the case when a member receives UNKNOWN_MEMBER_ID or FENCED_MEMBER_EPOCH from the leader. I'm clearing the memberId mainly for consistency, to avoid keeping it in-memory when we know is wrong. 

Just to double check with @dajac, in the case of these 2 errors, the member should rejoin (providing a member epoch 0), but I'm assuming there is no need to pass the previous member ID.

Reply:
i think the kip says we need to rejoin with the same id : `Upon receiving the UNKNOWN_MEMBER_ID or FENCED_MEMBER_EPOCH error, the consumer abandon all its partitions and rejoins with the same member id and the epoch 0.`

Reply:
Yes, I know it's not written like that but we had conversations saying that sending epoch 0 would be enough (that's why I want to double check)

Reply:
In general, we want to keep the member id but we have to clear it when receiving UNKNOWN_MEMBER_ID because as you said it is invalid.

Reply:
would be good to clarify that on the KIP! @dajac

Reply:
Actually, we could also keep the member id upon receiving UNKNOWN_MEMBER_ID. The coordinator will use it to recreate the new member. I think that it depends on whether we really want the ability to detect restarts of the consumer or not. If we really want to, keeping the member id would be better. What do you think?

Reply:
+1 on keeping it even in the UNKNOWN_MEMBER_ID scenario. Having the ability to differentiate restarts from brand new consumers seems to open the door for smarter behaviour.

Reply:
it is not entirely clear to me when would a client receive the UNKNOWN_MEMBER_ID error - Obviously, it could be caused by client side bug of sending the wrong/invalid memberId.  On the coordinator side, once a member is assigned to an id - I assume this shouldn't happen? Maybe when a group is being recreated or something?

Reply:
The server sends an UNKNOWN_MEMBER_ID error when the received member id does not exist and the received member epoch is greater than 0. This would happen if the member is fenced or manually removed from the consumer group.